File: llm_generated_code/ChatGPT/Python/Level1/Q5.py

def find_peak_element(nums):
    low, high = 0, len(nums) - 1

    while low < high:
        mid = (low + high) // 2
        if nums[mid] > nums[mid + 1]:
            # This means the peak is on the left side including mid
            high = mid
        else:
            # This means the peak is on the right side excluding mid
            low = mid + 1

    return low  # low and high will eventually converge

File: llm_generated_code/ChatGPT/CPP/Level1/Q5.cpp
#include <iostream>
#include <vector>
using namespace std;

int findPeakElement(const vector<int>& arr) {
    int left = 0;
    int right = arr.size() - 1;

    while (left < right) {
        int mid = left + (right - left) / 2;

        // If the middle element is less than its next element,
        // then the peak must be in the right half
        if (arr[mid] < arr[mid + 1]) {
            left = mid + 1;
        }
        // Otherwise, the peak is in the left half or could be the middle element itself
        else {
            right = mid;
        }
    }

    // At the end of the while loop, left and right will converge to the peak element index
    return left;
}

int main() {
    vector<int> arr = {1, 3, 20, 4, 1, 0}; // Example array
    int peakIndex = findPeakElement(arr);
    cout << "Index of a peak element: " << peakIndex << endl;
    return 0;
}

